// @number-flow/vue/dist/index.js
import { defineComponent as M, provide as F, renderSlot as L, watch as N, nextTick as T, onUnmounted as j, ref as v, computed as y, inject as V, openBlock as X, createElementBlock as G, mergeProps as z, unref as P, onMounted as H, watchEffect as R, toValue as x } from "vue";
import o, { formatToData as U, renderInnerHTML as W, define as I, canAnimate as Y, prefersReducedMotion as b } from "number-flow/lite";
import { default as le } from "number-flow/lite";
import { BROWSER as h } from "esm-env";
export * from "number-flow/plugins";
const k = String.raw;
h && (() => {
  try {
    document.createElement("div").animate({ opacity: 0 }, { easing: "linear(0, 1)" });
  } catch {
    return !1;
  }
  return !0;
})();
h && typeof CSS < "u" && CSS.supports && CSS.supports("line-height", "mod(1,1)");
h && typeof matchMedia < "u" && matchMedia("(prefers-reduced-motion: reduce)");
const _ = "--_number-flow-d-opacity", C = "--_number-flow-d-width", w = "--_number-flow-dx", E = "--_number-flow-d";
(() => {
  try {
    return CSS.registerProperty({
      name: _,
      syntax: "<number>",
      inherits: !1,
      initialValue: "0"
    }), CSS.registerProperty({
      name: w,
      syntax: "<length>",
      inherits: !0,
      initialValue: "0px"
    }), CSS.registerProperty({
      name: C,
      syntax: "<number>",
      inherits: !1,
      initialValue: "0"
    }), CSS.registerProperty({
      name: E,
      syntax: "<number>",
      inherits: !0,
      initialValue: "0"
    }), !0;
  } catch {
    return !1;
  }
})();
const S = "var(--number-flow-char-height, 1em)", g = "round(nearest, calc(var(--number-flow-mask-height, 0.25em) / 2), 1px)", s = `calc(${g} * 2)`, $ = "var(--number-flow-mask-width, 0.5em)", c = `calc(${$} / var(--scale-x))`, f = "#000 0, transparent 71%", D = k`:host{display:inline-block;direction:ltr;white-space:nowrap;isolation:isolate;line-height:${S} !important}.number,.number__inner{display:inline-block;transform-origin:left top}:host([data-will-change]) :is(.number,.number__inner,.section,.digit,.digit__num,.symbol){will-change:transform}.number{--scale-x:calc(1 + var(${C}) / var(--width));transform:translateX(var(${w})) scaleX(var(--scale-x));margin:0 calc(-1 * ${$});position:relative;-webkit-mask-image:linear-gradient(to right,transparent 0,#000 ${c},#000 calc(100% - ${c}),transparent ),linear-gradient(to bottom,transparent 0,#000 ${s},#000 calc(100% - ${s}),transparent 100% ),radial-gradient(at bottom right,${f}),radial-gradient(at bottom left,${f}),radial-gradient(at top left,${f}),radial-gradient(at top right,${f});-webkit-mask-size:100% calc(100% - ${s} * 2),calc(100% - ${c} * 2) 100%,${c} ${s},${c} ${s},${c} ${s},${c} ${s};-webkit-mask-position:center,center,top left,top right,bottom right,bottom left;-webkit-mask-repeat:no-repeat}.number__inner{padding:${g} ${$};transform:scaleX(calc(1 / var(--scale-x))) translateX(calc(-1 * var(${w})))}:host > :not(.number){z-index:5}.section,.symbol{display:inline-block;position:relative;isolation:isolate}.section::after{content:'\200b';display:inline-block}.section--justify-left{transform-origin:center left}.section--justify-right{transform-origin:center right}.section > [inert],.symbol > [inert]{margin:0 !important;position:absolute !important;z-index:-1}.digit{display:inline-block;position:relative;--c:var(--current) + var(${E})}.digit__num,.number .section::after{padding:${g} 0}.digit__num{display:inline-block;--offset-raw:mod(var(--length) + var(--n) - mod(var(--c),var(--length)),var(--length));--offset:calc( var(--offset-raw) - var(--length) * round(down,var(--offset-raw) / (var(--length) / 2),1) );--y:clamp(-100%,var(--offset) * 100%,100%);transform:translateY(var(--y))}.digit__num[inert]{position:absolute;top:0;left:50%;transform:translateX(-50%) translateY(var(--y))}.digit:not(.is-spinning) .digit__num[inert]{display:none}.symbol__value{display:inline-block;mix-blend-mode:plus-lighter;white-space:pre}.section--justify-left .symbol > [inert]{left:0}.section--justify-right .symbol > [inert]{right:0}.animate-presence{opacity:calc(1 + var(${_}))}`, O = k`:host{display:inline-block;direction:ltr;white-space:nowrap;line-height:${S} !important}span{display:inline-block}:host([data-will-change]) span{will-change:transform}.number,.digit{padding:${g} 0}.symbol{white-space:pre}`, q = (e = "") => k`number-flow${e} > span{font-kerning:none;display:inline-block;line-height:${S} !important;padding:${s} 0}`, J = (e) => [O, q(e), D], B = Symbol(), te = /* @__PURE__ */ M({
  __name: "NumberFlowGroup",
  setup(e) {
    const l = /* @__PURE__ */ new Set();
    let a = !1;
    return F(B, (i, m) => {
      l.add(i), N(
        m,
        async () => {
          a || (a = !0, l.forEach(async (t) => {
            var d;
            !t.value || !t.value.created || (t.value.willUpdate(), await T(), (d = t.value) == null || d.didUpdate());
          }), await T(), a = !1);
        }
        // { flush: 'pre' } // default
      ), j(() => {
        l.delete(i);
      });
    }), (i, m) => L(i.$slots, "default");
  }
}), K = ["batched", "trend", "plugins", "animated", "transformTiming", "spinTiming", "opacityTiming", "respectMotionPreference", "nonce", "data-will-change", "digits", "innerHTML", "data"], ne = /* @__PURE__ */ M({
  inheritAttrs: !1,
  __name: "index",
  props: {
    transformTiming: { default: () => o.defaultProps.transformTiming },
    spinTiming: { default: () => o.defaultProps.spinTiming },
    opacityTiming: { default: () => o.defaultProps.opacityTiming },
    animated: { type: Boolean, default: () => o.defaultProps.animated },
    respectMotionPreference: { type: Boolean, default: () => o.defaultProps.respectMotionPreference },
    trend: { type: [Number, Function], default: () => o.defaultProps.trend },
    plugins: { default: () => o.defaultProps.plugins },
    digits: { default: () => o.defaultProps.digits },
    locales: {},
    format: {},
    value: {},
    prefix: {},
    suffix: {},
    nonce: {},
    willChange: { type: Boolean, default: !1 }
  },
  emits: ["animationsstart", "animationsfinish"],
  setup(e, { expose: l, emit: a }) {
    const r = v();
    l({ el: r });
    const i = a, m = y(() => new Intl.NumberFormat(e.locales, e.format)), t = y(() => U(e.value, m.value, e.prefix, e.suffix)), d = h ? void 0 : W(t.value, { nonce: e.nonce, elementSuffix: "-vue" }), u = V(B, void 0);
    return u == null || u(r, t), (n, p) => (X(), G("number-flow-vue", z({
      ref_key: "el",
      ref: r
    }, n.$attrs, {
      batched: !!P(u),
      trend: n.trend,
      plugins: n.plugins,
      animated: n.animated,
      transformTiming: n.transformTiming,
      spinTiming: n.spinTiming,
      opacityTiming: n.opacityTiming,
      respectMotionPreference: n.respectMotionPreference,
      nonce: n.nonce,
      "data-will-change": n.willChange ? "" : void 0,
      digits: n.digits,
      innerHTML: P(d),
      "data-allow-mismatch": "",
      onAnimationsstart: p[0] || (p[0] = (A) => i("animationsstart")),
      onAnimationsfinish: p[1] || (p[1] = (A) => i("animationsfinish")),
      data: t.value
    }), null, 16, K));
  }
}), ie = J("-vue");
I("number-flow-vue", o);
function ae({
  respectMotionPreference: e = !0
} = {}) {
  const l = v(!1), a = v(!1);
  return H(() => {
    var r;
    l.value = Y, a.value = ((r = b) == null ? void 0 : r.matches) ?? !1;
  }), R((r) => {
    var m;
    if (!x(e)) return;
    const i = ({ matches: t }) => {
      a.value = t;
    };
    (m = b) == null || m.addEventListener("change", i), r(() => {
      var t;
      (t = b) == null || t.removeEventListener("change", i);
    });
  }), y(
    () => l.value && (!x(e) || !a.value)
  );
}
export {
  le as NumberFlowElement,
  te as NumberFlowGroup,
  ne as default,
  ie as styles,
  ae as useCanAnimate
};
